object-client/src/main/java/software/amazon/s3/analyticsaccelerator/S3SdkObjectClient.java

```
@@ -193,7 +201,20 @@ public CompletableFuture<ObjectContent> getObject(
         s3AsyncClient
             .getObject(builder.build(), AsyncResponseTransformer.toBlockingInputStream())
             .thenApply(
-                responseInputStream ->
-                    ObjectContent.builder().stream(responseInputStream).build()));
+                responseInputStream -> ObjectContent.builder().stream(responseInputStream).build())
+            .exceptionally(handleException(getRequest.getS3Uri())));
+  }
+
+  private <T> Function<Throwable, T> handleException(S3URI s3Uri) {
+    return throwable -> {
+      Throwable cause =
+          Optional.ofNullable(throwable.getCause())
+              .filter(
+                  t ->
+                      throwable instanceof CompletionException
+                          || throwable instanceof ExecutionException)
+              .orElse(throwable);
+      throw new UncheckedIOException(ExceptionHandler.toIOException(cause, s3Uri));
```

reading this again, i got confused why are we wrapping the IOException with UncheckedIOException. Why do we need it?

---

So, my thought process is this:
 
 - `handleException` method is used in a `CompletableFuture` chain, specifically in the `exceptionally` method above. Since `exceptionally` is a functional style, it doesn't allow the direct throwing of checked exceptions like `IOException.` 
 - By wrapping it in `UncheckedIOException,` we can propagate the exception in a way compatible with the functional-style code while deferring handling to higher-level logic. 
 - I felt this is cleaner than alternatives such as throwing as `RuntimeException` or moving to imperative style using try-catch.
 
Please let me know if you have any suggestions, and correct me if I'm wrong.